### ISSUE TITLE
`xqueue_file` field from `xqueue_body` response is not a valid download link

### ISSUE THREAD

**Issue by iammaxence** [overhangio/tutor-xqueue#2]
**Description of the issue :**

In the Edx documentation, they said that xqueue_files is a key/value pair where "each key is a file name, and each value is the location of the file?". 
They gave this example :
``` "xqueue_files": { "helloworld.c": "http://download.location.com/helloworld.c" } ```

In my case, instead of a link, i have the name of the docker folder where the files are stored. Like this : 
```


{

"xqueue_header" : {"submission_id": 1, "submission_key": "6cb73ff9c2e65c5bjk4cd6ce3aff6801"}, 
"xqueue_body" : { 
                  "grader_payload ": {"gitcourse": "exercice1"} ,
                  "xqueue_file" : {
                                   "myfile.zip": "openedx/1574f4c2eca142d0d7da68f1b9111e86"
                                   }
                 }
}

```
Does Tutor allow some configuration to have a direct download link (like a endpoint to get the file easily) ?

link of the discussion on the forum : https://discuss.overhang.io/t/tutor-xqueue-plugin-xqueue-files-link-to-download/1554

**How to possibly solve this issue :** 

In the xqueue/settings.py from the Xqueue repo of Edx, there is a MEDIA_URL and MEDIA_ROOT variables:

1. MEDIA_ROOT is the path where the submit files are upload.
2. MEDIA_URL to access the files that are stored in MEDIA_ROOT

We can add those variables in `tutorxqueue/templates/xqueue/apps/settings/tutor.py` like this : 
```
MEDIA_ROOT = '/edx/files_submission/'

MEDIA_URL = '/files_submission/'
```
Now, we are able to see something like this : 

```
{

"xqueue_header" : {"submission_id": 1, "submission_key": "6cb73ff9c2e65c5bjk4cd6ce3aff6801"}, 
"xqueue_body" : { 
                  "grader_payload ": {"gitcourse": "exercice1"} ,
                  "xqueue_file" : {
                                   "myfile.zip": "/files_submission/openedx/1574f4c2eca142d0d7da68f1b9111e86"
                                   }
                 }
}
```
To be able to get the file, i think we need to modify the urlpatterns from `submission_queue/urls.py` from the Edx XQueue repo.

Change this : 
```
# External pulling interface
# ------------------------------------------------------------
urlpatterns += [
    url(r'^get_queuelen/$', get_queuelen),
    url(r'^get_submission/$', get_submission),
    url(r'^put_result/$', put_result),
]
```
To this : 

```
urlpatterns += [
    url(r'^get_queuelen/$', get_queuelen),
    url(r'^get_submission/$', get_submission),
    url(r'^put_result/$', put_result),
] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```
Do i need to create an issue in the XQueue repo of Edx or can we do otherwise ?

**Comment by regisb** [comment]
I am not sure yet whether this warrants an upstream change. Let's wait until someone has time to prioritize this issue and actually have a look at it.

**Comment by alfredchavez** [comment]
Hi @regisb, I have been checking on xqueue and that problem with media files, I think that xqueue is not supposed to provide a direct download link(so far) if the default storage is local as this "problem" is present even if media root and media url are configured and without tutor intervention.
The proposed solution to serve media files seems great, but it is not recommended as it is for development purposes(also needs DEBUG=True) https://docs.djangoproject.com/en/3.0/howto/static-files/#serving-static-files-during-development  and I have seen that it is not encouraged to use django to serve this media files(better use apache/nginx conf or cloud providers like aws s3).
I don't think it may need a change in edx/xqueue or tutor-xqueue.

**Comment by regisb** [comment]
@alfredchavez thanks for investigating this! Did you figure out what's the actual content of these media files? For instance, you can exec inside the xqueue container to check the media folder and the contents of the media files.

@iammaxence can you please clarify what you need these media files for? I wasn't even aware that xqueue needed to store media files somewhere.

**Comment by iammaxence** [comment]
> I think that xqueue is not supposed to provide a direct download link

That's weird.. How people are supposed to use External Graders with xqueue if they can't reach the submission file properly? 

> For instance, you can exec inside the xqueue container to check the media folder and the contents of the media files.

The media folder contains all the submissions files from an External Grader exercise.

> can you please clarify what you need these media files for? I wasn't even aware that xqueue needed to store media files somewhere.

@regisb I would like to make an auto-correct exercise using Externals Graders. For example, someone make an algorithm in C++ then submit the file for correction. I need this file to be able to give a correction to the user. By default, the file is stored in the xqueue container.

For now, i manage to get the file with a **docker cp** (I copy the file from the container). But containers are ephemeral..If my container restart, i will loose the files from submissions.

**Comment by alfredchavez** [comment]
> That's weird.. How people are supposed to use External Graders with xqueue if they can't reach the submission file properly?

 I have seen this problem if the storage is "local"(in the container itself), but if an external cloud provider like s3 is set as the media files default storage, the download link is going to be correctly provided(pointing to the file location in the s3 bucket)
>Did you figure out what's the actual content of these media files? For instance, you can exec inside the xqueue container to check the media folder and the contents of the media files.

As mentioned by @iammaxence , the media folder contains the submission files in case default storage is set to save them locally, if default storage is used with boto3 and setup with s3 is going to save the files on the s3 bucket

**Comment by regisb** [comment]
I'm not going to have time to look at this for a couple weeks. Anyone interested @overhangio/tutor-developers?

**Comment by regisb** [comment]
Thanks for the report @alfredchavez, and sorry for the delay in processing this. The fix is now included in 12.1.0. The response of `tutor submissions show` now includes a link to the uploaded files. I am also going to publish an update to the minio plugin in order to support object storage, which is necessary when deploying to Kubernetes.
